### PR TITLE
Extract and restore a test that a blank native assets project still builds

### LINKED ISSUE #158120: Failing postsubmit test: 'FormatException: No value was provided for required key: out_dir_shared'
### Type of Request

bug

### Infrastructure Environment

Postsubmit tests (LUCI)

### What is happening?

<details>
<summary>Error logs</summary>

```
05:25 +30: test/integration.shard/isolated/native_assets_test.dart: flutter build apk succeeds without libraries
05:38 +30 -1: test/integration.shard/isolated/native_assets_test.dart: flutter build apk succeeds without libraries [E]
  Exception: flutter build failed: 1
  Unhandled exception:
  FormatException: No value was provided for required key: out_dir_shared
  #0      MapJsonUtils.get (package:native_assets_cli/src/json_utils.dart:50:7)
  #1      MapJsonUtils.path (package:native_assets_cli/src/json_utils.dart:25:48)
  #2      new HookConfig (package:native_assets_cli/src/config.dart:88:38)
  #3      new BuildConfig (package:native_assets_cli/src/config.dart)
  #4      build (package:native_assets_cli/src/api/build.dart:99:18)
  #5      main (file:///Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/hook/build.dart:4:9)
  #6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
  #7      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
  Building native assets for package:package_with_native_assets failed.
  file:///Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/hook/build.dart returned with exit code: 255.
  To reproduce run:
  /Volumes/Work/s/w/ir/x/w/flutter/bin/dart --packages=/Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/.dart_tool/package_config.json /Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/.dart_tool/native_assets_builder/545fcee0f93c80dfaac63b71612fc26c/hook.dill --config=/Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/.dart_tool/native_assets_builder/545fcee0f93c80dfaac63b71612fc26c/config.json
  stderr:
  Unhandled exception:
  FormatException: No value was provided for required key: out_dir_shared
  #0      MapJsonUtils.get (package:native_assets_cli/src/json_utils.dart:50:7)
  #1      MapJsonUtils.path (package:native_assets_cli/src/json_utils.dart:25:48)
  #2      new HookConfig (package:native_assets_cli/src/config.dart:88:38)
  #3      new BuildConfig (package:native_assets_cli/src/config.dart)
  #4      build (package:native_assets_cli/src/api/build.dart:99:18)
  #5      main (file:///Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/hook/build.dart:4:9)
  #6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
  #7      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
  
  stdout:
  
  
  Target native_assets failed: Error: Building native assets failed. See the logs for more details.
  
  
  FAILURE: Build failed with an exception.
  
  * What went wrong:
  Execution failed for task ':app:compileFlutterBuildDebug'.
  > Process 'command '/Volumes/Work/s/w/ir/x/w/flutter/bin/flutter'' finished with non-zero exit value 1
  
  * Try:
  > Run with --stacktrace option to get the stack trace.
  > Run with --info or --debug option to get more log output.
  > Run with --scan to get full insights.
  > Get more help at https://help.gradle.org.
  
  BUILD FAILED in 9s
  Gradle task assembleDebug failed with exit code 1
  
  Resolving dependencies...
  Downloading packages...
    async 2.11.0 (2.12.0 available)
    boolean_selector 2.1.1 (2.1.2 available)
    characters 1.3.0 (1.3.1 available)
    clock 1.1.1 (1.1.2 available)
    collection 1.19.0 (1.19.1 available)
    fake_async 1.3.1 (1.3.2 available)
    matcher 0.12.16+1 (0.12.17 available)
    material_color_utilities 0.11.1 (0.12.0 available)
    meta 1.15.0 (1.16.0 available)
    path 1.9.0 (1.9.1 available)
    string_scanner 1.3.0 (1.4.0 available)
    vm_service 14.3.0 (14.3.1 available)
  Got dependencies!
  12 packages have newer versions incompatible with dependency constraints.
  Try `flutter pub outdated` for more information.
  
  Running Gradle task 'assembleDebug'...                          
  Running Gradle task 'assembleDebug'...                             10.2s
  
  test/integration.shard/isolated/native_assets_test.dart 241:11       main.<fn>.<fn>
  ===== asynchronous gap ===========================
  test/integration.shard/isolated/native_assets_test_utils.dart 275:5  inTempDir
  ===== asynchronous gap ===========================
  test/integration.shard/isolated/native_assets_test.dart 227:7        main.<fn>
```
</details>

### Steps to reproduce

_No response_

### Expected results

_No response_

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/flutter/flutter/issues/158120.

This PR restores the skipped test, moving it (and the test utility only used by the test) into a standalone file that can be more easily understood. As part of the change the version of `native_assets_cli` is now derived from the (checked-in) `package_ffi/pubspec.yaml.tmpl`, meaning that it should be hard to get into a bad state again.

/cc @christopherfujino (You are welcome to review, but otherwise will defer to Brandon and Victor).

### CODE REVIEW

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+/// Reads `templates/package_ffi/pubspec.yaml.tmpl` to use the package version.
+///
+/// For example, if the template would output:
+/// ```yaml
+/// dependencies:
+///   native_assets_cli: ^0.8.0
+/// ```
+///
+/// ... theh this function would return `'^0.8.0'`.
```
Comment (victorsanni) [inline packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart:139]:
```suggestion
/// ... then this function would return `'^0.8.0'`.
```

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+String _getPackageFfiTemplatePubspecVersion() {
+  final String path = Context().join(
+    getFlutterRoot(),
+    'packages',
+    'flutter_tools',
+    'templates',
+    'package_ffi',
```
Comment (bdero) [inline packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart:146]: Just to note: This test isn't FFI-related, but I like that we can share the test version pin.

**Review by bdero — APPROVED** [review]
LGTM, thanks!

**Review by victorsanni — APPROVED** [review]
LGTM with the caveat that I know nothing about flutter tools/native assets/pub.

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart (hunk 2 above)
Comment (matanlurey) [reply packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart:146]: Agreed!

**Comment by matanlurey** [comment]
> LGTM with the caveat that I know nothing about flutter tools/native assets/pub.

No sweat, just wanted to give the gardener visibility into this potentially re-landing.

**Comment by christopherfujino** [comment]
FYI @bkonyi

**Comment by matanlurey** [comment]
> Failed to create CP due to merge conflicts.You will need to create the PR manually. See [the cherrypick wiki](https://github.com/flutter/flutter/blob/main/docs/releases/Flutter-Cherrypick-Process.md) for more info.

@reidbaker LMK if you need help manually cherry-picking/merging.

**Comment by reidbaker** [comment]
I can do it but this is like 4 or 5 things deep on my priority list. If someone else hits this issue feel free to add me as a reviewer to the 3.27 branch.

**Comment by matanlurey** [comment]
I am waiting for some help from the iOS team, so I can pick this up.

**Pull request merged**